webapp/packages/plugin-data-viewer/src/useDataViewerModel.ts

```diff
@@ -0,0 +1,119 @@
+/*
+ * CloudBeaver - Cloud Database Manager
+ * Copyright (C) 2020-2024 DBeaver Corp and others
+ *
+ * Licensed under the Apache License, Version 2.0.
+ * you may not use this file except in compliance with the License.
+ */
+import { action, computed, observable } from 'mobx';
+
+import { useObservableRef, useResource } from '@cloudbeaver/core-blocks';
+import { ConnectionInfoResource, IConnectionInfoParams } from '@cloudbeaver/core-connections';
+import { ILoadableState, isContainsException } from '@cloudbeaver/core-utils';
+import type { ITab } from '@cloudbeaver/plugin-navigation-tabs';
+
+import type { DataViewerTable } from './DataViewerTable';
+
+interface ITabState {
+  connectionKey: IConnectionInfoParams | undefined;
+  tableId?: string;
+}
+
+type Tab = ITab<ITabState>;
+
+interface State extends ILoadableState {
+  resetException(): void;
+}
+
+export interface IDataViewerDatabaseDataModel extends State {
+  dataViewerTable: DataViewerTable;
+  tab: Tab;
+  _exception?: Error[] | Error | null;
+  _loading: boolean;
+  _init(): Promise<void>;
+  init(tab: Tab, state: State): Promise<void>;
+  load(): Promise<void>;
+}
+
+export function useDataViewerModel<T extends Tab>(tab: T, dataViewerTable: DataViewerTable, init: (tab: T, state: State) => Promise<void>) {
+  const connection = useResource(useDataViewerModel, ConnectionInfoResource, tab.handlerState.connectionKey ?? null);
+
+  const state = useObservableRef<IDataViewerDatabaseDataModel>(
+    () => ({
+      _exception: null,
+      _loading: false,
+      get exception() {
+        if (isContainsException(connection.exception)) {
+          return connection.exception;
+        }
+
+        return this._exception;
+      },
+      isLoading(): boolean {
+        return connection.isLoading() || this._loading;
+      },
+      isLoaded(): boolean {
+        return connection.isLoaded() && this.dataViewerTable.get(this.tab.handlerState.tableId || '') !== undefined;
+      },
+      isError(): boolean {
+        return this.exception !== null;
```

Comment:
exception is also can be undefined

I believe something like this is more suitable here:
`return isNotNullDefined(this.exception)`

Reply:
right, but as long as it implements ILoadableState it's better to use `isContainsException` cuz it accepts all variants of exception value